File: veinview/src/main/java/com/izikode/izilib/veinview/VeinView.java
package com.izikode.izilib.veinview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.RawRes;
import android.util.AttributeSet;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class VeinView extends WebView {

    public VeinView(Context context) {
        super(context);
        init();
    }

    public VeinView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public VeinView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    /**
     * @deprecated Use setVeinViewClient to get injection functionality
     */
    @Deprecated
    @Override
    public void setWebViewClient(WebViewClient client) {
        super.setWebViewClient(client);
    }

    public void setVeinViewClient(@NonNull VeinViewClient client) {
        client.setInjector(injector);
        super.setWebViewClient(client);
    }

    private String injectCSS, injectJS;
    private VeinViewInjector injector;

    @SuppressLint("SetJavaScriptEnabled")
    private void init() {
        injectCSS = resource(R.raw.inject_css);
        injectJS = resource(R.raw.inject_js);

        injector = new VeinViewInjector(new VeinViewInjections() {

            @Override
            public void css(@NonNull String code) {
                injectCSS(code);
            }

            @Override
            public void css(int codeResource) {
                injectCSS(codeResource);
            }

            @Override
            public void js(@NonNull String code) {
                injectJS(code);
            }

            @Override
            public void js(int codeResource) {
                injectJS(codeResource);
            }

        });

        getSettings().setJavaScriptEnabled(true);
    }

    private void injectCSS(@NonNull String css) {
        run(String.format(injectCSS, css.trim()));
    }

    private void injectCSS(@RawRes int cssResource) {
        injectCSS(resource(cssResource));
    }

    private void injectJS(@NonNull String js) {
        run(String.format(injectJS, js.trim()));
    }

    private void injectJS(@RawRes int jsResource) {
        injectJS(resource(jsResource));
    }

    private void run(@NonNull String javascript) {
        String code = String.format("javascript:(function() { %1$s })()", javascript);
        loadUrl(code);
    }

    private String resource(@RawRes int resource) {
        InputStream raw = getResources().openRawResource(resource);
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        try {

            int chunk = raw.read();

            while (chunk != -1) {
                out.write(chunk);
                chunk = raw.read();
            }

            raw.close();

        } catch (IOException e) {
            e.printStackTrace();
        }

        return out.toString();
    }

}

File: veinview/src/main/java/com/izikode/izilib/veinview/VeinView.kt
package com.izikode.izilib.veinview

import android.annotation.SuppressLint
import android.content.Context
import android.support.annotation.RawRes
import android.util.AttributeSet
import android.webkit.WebView
import android.webkit.WebViewClient
import java.io.ByteArrayOutputStream
import java.io.IOException

@SuppressLint("SetJavaScriptEnabled")
class VeinView : WebView {

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    private val injectCSS: String by lazy { resource(R.raw.inject_css) }
    private val injectJS: String by lazy { resource(R.raw.inject_js) }

    private val injector: VeinViewInjector by lazy {

        VeinViewInjector(object : VeinViewInjections {

            override fun css(code: String) {
                injectCSS(code)
            }

            override fun css(codeResource: Int) {
                injectCSS(codeResource)
            }

            override fun js(code: String) {
                injectJS(code)
            }

            override fun js(codeResource: Int) {
                injectJS(codeResource)
            }

        })
    }

    init {
        settings.javaScriptEnabled = true
    }

    @Deprecated("Use setVeinViewClient to get injection functionality",
            ReplaceWith("setVeinViewClient(client)", "android.webkit.WebView"))
    override fun setWebViewClient(client: WebViewClient) {
        super.setWebViewClient(client)
    }

    fun setVeinViewClient(client: VeinViewClient) {
        client.injector = injector
        super.setWebViewClient(client)
    }

    private fun injectCSS(css: String) {
        run(String.format(injectCSS, css.trim { it <= ' ' }))
    }

    private fun injectCSS(@RawRes cssResource: Int) {
        injectCSS(resource(cssResource))
    }

    private fun injectJS(js: String) {
        run(String.format(injectJS, js.trim { it <= ' ' }))
    }

    private fun injectJS(@RawRes jsResource: Int) {
        injectJS(resource(jsResource))
    }

    private fun run(javascript: String) {
        val code = "javascript:(function() { $javascript })()"
        loadUrl(code)
    }

    private fun resource(@RawRes resource: Int): String {
        val raw = resources.openRawResource(resource)
        val out = ByteArrayOutputStream()

        try {

            var chunk = raw.read()

            while (chunk != -1) {
                out.write(chunk)
                chunk = raw.read()
            }

            raw.close()

        } catch (e: IOException) {
            e.printStackTrace()
        }

        return out.toString()
    }

}
